File: ports/csharp/src/core/kernel/BcsCommandLine.cs
using System;
using System.Collections.Generic;

namespace Bcs.Core.Kernel
{
    public class BcsCommandLineOption
    {
        public List<string> Names { get; set; } = new List<string>();
        public string Description { get; set; } = string.Empty;
        public string ValueName { get; set; } = string.Empty;
        public string DefaultValue { get; set; } = string.Empty;
    }

    public class BcsCommandLineParser
    {
        private List<BcsCommandLineOption> _options = new List<BcsCommandLineOption>();
        private Dictionary<string, List<string>> _parsedOptions = new Dictionary<string, List<string>>();
        private List<string> _positionalArgs = new List<string>();
        private string _appDescription = string.Empty;

        public void AddOption(BcsCommandLineOption option)
        {
            _options.Add(option);
        }

        public void SetApplicationDescription(string desc)
        {
            _appDescription = desc;
        }

        public void AddHelpOption()
        {
            AddOption(new BcsCommandLineOption {
                Names = new List<string> { "h", "help" },
                Description = "Displays this help."
            });
        }

        public void AddVersionOption()
        {
            AddOption(new BcsCommandLineOption {
                Names = new List<string> { "v", "version" },
                Description = "Displays version information."
            });
        }

        public bool Parse(string[] args)
        {
            for (int i = 1; i < args.Length; i++)
            {
                var arg = args[i];
                if (arg == "-h" || arg == "--help")
                {
                    _parsedOptions["help"] = new List<string> { "true" };
                }
                else if (arg == "-v" || arg == "--version")
                {
                    _parsedOptions["version"] = new List<string> { "true" };
                }
                else
                {
                    _positionalArgs.Add(arg);
                }
            }
            return true;
        }

        public void Process(string[] args)
        {
            Parse(args);
            if (IsSet("help")) ShowHelp(0);
            if (IsSet("version")) ShowVersion();
        }

        public bool IsSet(string name)
        {
            return _parsedOptions.ContainsKey(name);
        }

        public void ShowHelp(int exitCode)
        {
            Console.WriteLine(_appDescription);
            Console.WriteLine("Options:");
            foreach (var opt in _options)
            {
                Console.WriteLine($"  -{opt.Names[0]}, --{opt.Names[opt.Names.Count - 1]} : {opt.Description}");
            }
            Environment.Exit(exitCode);
        }

        public void ShowVersion()
        {
            Console.WriteLine("Version 1.0.0");
            Environment.Exit(0);
        }
    }
}

File: ports/rust/src/core/kernel/bcs_commandline.rs
use std::collections::HashMap;

pub struct BcsCommandLineOption {
    pub names: Vec<String>,
    pub description: String,
    pub value_name: String,
    pub default_value: String,
}

pub struct BcsCommandLineParser {
    options: Vec<BcsCommandLineOption>,
    parsed_options: HashMap<String, Vec<String>>,
    positional_args: Vec<String>,
    app_description: String,
}

impl BcsCommandLineParser {
    pub fn new() -> Self {
        Self {
            options: Vec::new(),
            parsed_options: HashMap::new(),
            positional_args: Vec::new(),
            app_description: String::new(),
        }
    }

    pub fn add_option(&mut self, option: BcsCommandLineOption) -> bool {
        self.options.push(option);
        true
    }

    pub fn set_application_description(&mut self, desc: &str) {
        self.app_description = desc.to_string();
    }

    pub fn add_help_option(&mut self) {
        self.add_option(BcsCommandLineOption {
            names: vec!["h".to_string(), "help".to_string()],
            description: "Displays this help.".to_string(),
            value_name: String::new(),
            default_value: String::new(),
        });
    }

    pub fn add_version_option(&mut self) {
        self.add_option(BcsCommandLineOption {
            names: vec!["v".to_string(), "version".to_string()],
            description: "Displays version information.".to_string(),
            value_name: String::new(),
            default_value: String::new(),
        });
    }

    pub fn parse(&mut self, args: Vec<String>) -> bool {
        for arg in args.into_iter().skip(1) {
            if arg == "-h" || arg == "--help" {
                self.parsed_options.insert("help".to_string(), vec!["true".to_string()]);
            } else if arg == "-v" || arg == "--version" {
                self.parsed_options.insert("version".to_string(), vec!["true".to_string()]);
            } else {
                self.positional_args.push(arg);
            }
        }
        true
    }

    pub fn process(&mut self, args: Vec<String>) {
        self.parse(args);
        if self.is_set("help") {
            self.show_help(0);
        }
        if self.is_set("version") {
            self.show_version();
        }
    }

    pub fn is_set(&self, name: &str) -> bool {
        self.parsed_options.contains_key(name)
    }

    pub fn show_help(&self, exit_code: i32) {
        println!("{}", self.app_description);
        println!("Options:");
        for opt in &self.options {
            println!("  -{}, --{} : {}", opt.names[0], opt.names.last().unwrap(), opt.description);
        }
        std::process::exit(exit_code);
    }

    pub fn show_version(&self) {
        println!("Version 1.0.0");
        std::process::exit(0);
    }
}
